### PR TITLE
[ANDROID] [FIXED] Fixed random styling for text nodes with many children

### PR DESCRIPTION
**Pull request by cubuspl42**
Attempting to fix the issue https://github.com/facebook/react-native/issues/33418

## Changelog

[ANDROID] Fixed inconsistent styling for text nodes with many children

## Test Plan

No test plan yet. I'd like to ask for help with creating one.

##

Putting template aside, I'd like to ask for a review of the approach I'm suggesting.

React Native as-is (at least in some cases) [messes up the styles](https://github.com/facebook/react-native/issues/33418#issuecomment-1485305556) for text nodes with more than 85 children, just like that.

image

All of this text should be blue.

The root cause is that code (on Android) assumes it can assign as many `Spannable` span priority values as we'd like, while in reality, it has to be packed in an 8-bit-long section of the flags mask. So 255 possible values. In the scenario I produced, React generates three spans per text node, and for 85 text nodes, it sums up to 255. For each span, a new priority value is assigned.

As I understand it, we don't need that many priority values. If I'm not mistaken, these priorities are crucial only for ensuring that nested styles have precedence over the outer ones. I'm proposing to calculate the priority value "vertically" (based on the node's depth in the tree) not "horizontally" (based on its position).

It would be awesome if some core engineer familiar with `ReactAndroid` shared their experience with this module, especially if there are any known cases when we _know_ that we'd like to create overlapping spans fighting over the same aspects of the style.

### CODE REVIEW

**Comment by cubuspl42** [comment]
This branch is based on 0.71.4 because:

- That's the version I'm using
- [The "How to build from source" docs](https://reactnative.dev/contributing/how-to-build-from-source) just don't work on `main`

That having said, the bug is [still present on `main`](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactBaseTextShadowNode.java#L91).

We can discuss rebasing this these changes on `main` later, focusing on the big-picture stuff first.

**Comment by NickGerleman** [comment]
I am currently on vacation but can give some tentative feedback.

1. Overall change of relying on depth for priority makes sense
1. ReactBaseTextShadowNode is specific to Paper. The same change would need to also be made to TextLayoutManager and TextLayoutManagerMapBuffer to work in Fabric (that duplication will eventually be cleaned up)
1. Some spans may read metrics like ascent or descent height. We should check that using the same priority per depth leads to an ordering where these still get the right metrics.
2. Instead of hard-coding max number of spans, we should use SPAN_PRIORITY and SPAN_PRIORITY_SHIFT to derive this (e.g. in case it changes in the future).

**Comment by cubuspl42** [comment]
@NickGerleman Thanks for the early review! That helps a lot.

Ad. 2. I know about the ongoing Fabric thing but as a downstream React Native 0.71.4 user, that's a song of the future for me 😀 I'll figure that out and will try to apply the changes there too. But I think I would need some help with building the `main` branch. That might be something trivial, but I just didn't have much luck there yet. I'll share build failure messages when/if I encounter them again.

Ad 3. That sounds like a great thing to check and I'd like to check that. But, to be honest, I don't fully understand how that translates to React components terms. Would you be able to share one example of a `<Text>` component where this ascent/descent height reading might happen on the native side?

Ad 4. You're right, that should be possible.

By the way, do you know any way to calm the bot down?

**Comment by Szymon20000** [comment]
I checked the pr and it makes a lot of sense. 
Also tested it in RNTester on paper arch and wasn't able to spot any difference. 
As I'm not the author of the original code so I'm not able to tell what was the reason for using priorities this way. 
However, I guess the purpose was to make sure that children operations are executed after parents. 
I don't know any example of span operations that would be broken by this approach. 

**I'm not sure though if we really need to use priority. From what I see spans are sorted first by priority then by insertion order. I guess it wasn't the case when the code was created.** 

https://github.com/aosp-mirror/platform_frameworks_base/commit/fa05ba0b0d39fae1d2cb3d98fbee0aef6a9fed88
Screenshot 2023-04-03 at 17 51 52

Maybe we can set the same priority for span operations and insertion order will handle the rest (we also need to call Collections.reverse(ops) as we want to apply parents first)?

**Comment by cubuspl42** [comment]
@Szymon20000 I started from user messages containing hundreds of markdown tags, went through debugging the HTML library, then entered React Native, then `ReactAndroid`.

I think I reached my limit. 😅

I don't know if this sorting can be relied on. Does it happen on all Android versions supported by React Native? Is it guaranteed that it won't change (again?) in the future?

It would be great to have a quote from the API documentation that at least implies that the order actually matters. Otherwise, I don't know. I'm not sure if dropping priorities altogether is a good idea.

**Comment by Szymon20000** [comment]
Hard to check unfortunately. I think we can do a hybrid approach though.
assign priorities according to depth but also reverse the ops list. That should behave as priorities now. But without 255 limitation and no risk.

**Comment by cubuspl42** [comment]
@NickGerleman What do you think about comments by @Szymon20000?

**Comment by NickGerleman** [comment]
That makes sense on first glance, though I am not aware of how ordering is preserved when applying spans to the underlying TextPaint.

So, I’d want to see validation of the case where spans do influence each other. E.g. AbsoluteSizeSpan may influence the metrics read by CustomLineHeightSpan.

Then the other, was just making sure this logic is updated everywhere. Right now the various usages of “SetSpanOperation” are copy/pasted in a few places, but that should make it mechanical to validate the change in one place then apply to the others.

**Comment by cubuspl42** [comment]
@NickGerleman Ok, so I understand now that these action points are on me:
- Provide an example that verifies that spans that potentially influence each other work correctly after my PR
- Copy/paste the solution to places with this duplicated logic; the de-duplication is out of scope

Also, I guess you would expect the changes to be applied on `main`? Would you provide me some guidance on how can I point a React Native app to use my forked version built from source? [These docs](https://reactnative.dev/contributing/how-to-build-from-source#update-your-project-to-build-from-source) seem to be outdated and I had trouble figuring out how to make things build correctly.

**Comment by NickGerleman** [comment]
I think the most common workflow building RNTester in the repo and testing with JavaScript code there.

cc @cortinico re docs for building from source

**Comment by fabriziobertoglio1987** [comment]
https://github.com/facebook/react-native/pull/36656#issuecomment-1514677350
> So, I’d want to see validation of the case where spans do influence each other. E.g. AbsoluteSizeSpan may influence the metrics read by CustomLineHeightSpan.

Case 1:
```javascript
<Text>Parent Text
   <Text style={{fontSize:25, lineHeight: 30}}>Child Text</Text>
</Text>
```

Case 2:
```javascript
<Text>Parent Text
   <Text style={{lineHeight: 60}}>Child Text</Text>
   <Text style={{fontSize:25}}>Child Text</Text>
</Text>
```


Related to https://github.com/Expensify/App/issues/15648#issuecomment-1482386572. There are some info on this API in the discussions https://github.com/facebook/react-native/pull/35949, https://github.com/facebook/react-native/pull/35704

**Comment by fabriziobertoglio1987** [comment]
1) [lineHeight](https://github.com/fabriziobertoglio1987/react-native/blob/9baf641cac70b8473e6afdcead91046ff7758041/ReactAndroid/src/main/java/com/facebook/react/views/text/CustomLineHeightSpan.java#L17) is applied using [ParagraphStyle](https://github.com/aosp-mirror/platform_frameworks_base/blob/405a99e5aeb47751701be0d4e179718dd7ccee2f/core/java/android/text/style/LineHeightSpan.java#L33) span. You can change lineHeight of a paragraph, but not a character
2) fontSize is applied using [CharactherStyleSpan](https://github.com/aosp-mirror/platform_frameworks_base/blob/405a99e5aeb47751701be0d4e179718dd7ccee2f/core/java/android/text/style/AbsoluteSizeSpan.java#L35). You can change fontSize of each character.

More info from from https://github.com/facebook/react-native/pull/35704

MetricAffectingSpan allows to vertically align text (group of characters, not paragraphs divided by `\n`) by changing the TextPaint baseline of a span.

>Metric Affecting spans. Spans can also affect text metrics, such as line height and size. All of these spans extend the [MetricAffectingSpan](https://developer.android.com/reference/android/text/style/MetricAffectingSpan) class.

Text is represented as TextView on Android, Nested Text is represented as Spans.
https://developer.android.com/develop/ui/views/text-and-emoji/spans

https://developer.android.com/develop/ui/views/text-and-emoji/spans#span-types
>- How the span affects text: a span can affect either text appearance or text metrics.
>- Span scope: some spans can be applied to individual characters, while others must be applied to an entire paragraph

They consist of four types:
>- Paragraph Affecting spans (text separated by `\n` (new line)). A span can also affect text at a paragraph level, such as changing the alignment or the margin of the entire block of text. Spans that affect entire paragraphs implement [ParagraphStyle](https://developer.android.com/reference/android/text/style/ParagraphStyle) (.
>- Character Affecting spans. A span can affect text at a character level. For example, you can update character elements like background color, style, or size. Spans that affect individual characters extend the [CharacterStyle](https://developer.android.com/reference/android/text/style/CharacterStyle) class.
>- Appearance Affecting spans
>- Metric Affecting spans. Spans can also affect text metrics, such as line height and text size. All of these spans extend the [MetricAffectingSpan](https://developer.android.com/reference/android/text/style/MetricAffectingSpan) class.

**Comment by cortinico** [comment]
> [These docs](https://reactnative.dev/contributing/how-to-build-from-source#update-your-project-to-build-from-source) seem to be outdated and I had trouble figuring out how to make things build correctly.

@cubuspl42 those docs are actually the correct one. They work fine as long as you're on the latest 0.71+ patch

**Comment by cubuspl42** [comment]
@fabriziobertoglio1987 Thank you for all the input and quotes!

To be absolutely honest, I didn't understand the conclusion in the context of this PR. So are you saying that the appropriate tests that should prove that this PR does not introduce a regression are present in `CustomStyleSpanTest.java`, or that they could be added there, or neither?

**Comment by fabriziobertoglio1987** [comment]
The Java tests (AOSP, ReactAndroid) are useful as API documentation. I wanted to further elaborate on the above discussion, but did not reach a conclusion. My next step will be to look into the AOSP Span Tests and ReactAndroid Tests, as they have additional info on the Android API.

I think your changes are not specific to CustomStyleSpan. I tried to write similar integration tests. I think the test needs to use UIManager to add a ReactTextView. You can find similar test in ReactAndroid.

- add a ReactCustomLineHeight span to the ReactTextView
- add AbsoluteSizeSpan to ReactTextView 
- run some assertions 

I tried to write similar tests in the past, but failed to do that, but I found it easier to understand ReactAndroid by reading and writing the tests rather than reading the codebase.

I'm replying on mobile, I will write later a more complete answer.

**Comment by fabriziobertoglio1987** [comment]
- Use reactTextViewManager createViewInstance to create instance of ReactTextView
- Mock Yoga with PowerMockito. ReactTextView uses yoga and the test will fail if Yoga is not mocked. Mocking YogaConfigFactory.create and stubbing all not relevant methods could solve this issue.

As explained above, I was not able to write the test and I dont suggest you to do that, but the documentation included in this tests is very good.

Sorry for not clear reply, I'm on a mobile device.

Below my notes on this task.

> 
> 
> P | type | task
> -- | -- | --
> 8 | refactor | Unit Tests for TtsSpan (link, addRootView, ReactShadowNodeImpl, SimpleViewTest, createMockCatalystInstance, ReactInstrumentationTest)
> 8.1 | task | Mock YogaConfigFactory
> 8.2 | task | Read runtime stack of test (at com.facebook.yoga.YogaConfigFactory.create(YogaConfigFactory.java:12))
> 8.3 | task | Read Test cases YogaLayoutDirection. They also use YogaConfigFactory.create()
> 8.4 | task | Write a simple test to verify correct behaviour with specific accLabel and accSpan
> 8.5 | task | Use reactTextViewManager createViewInstance to create instance of ReactTextView
> 8.6 | task | Test: reactTextViewManager updateState adds a TtsSpan to text
> 
> P	type	task
> 8	refactor	Unit Tests for TtsSpan ([link](https://github.com/facebook/react-native-deprecated-modules/pull/16#issuecomment-1423088469), [addRootView](https://github.com/fabriziobertoglio1987/react-native/blob/d536cf73fd8579844a17e8676aade9b2a92df0d6/ReactAndroid/src/test/java/com/facebook/react/views/text/ReactTextTest.java#L478), [ReactShadowNodeImpl](https://github.com/fabriziobertoglio1987/react-native/blob/d536cf73fd8579844a17e8676aade9b2a92df0d6/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactShadowNodeImpl.java#L62), [SimpleViewTest](https://github.com/fabriziobertoglio1987/react-native/blob/d536cf73fd8579844a17e8676aade9b2a92df0d6/ReactAndroid/src/test/java/com/facebook/react/uimanager/SimpleViewPropertyTest.java#L81), [createMockCatalystInstance](https://github.com/fabriziobertoglio1987/react-native/blob/d536cf73fd8579844a17e8676aade9b2a92df0d6/ReactAndroid/src/test/java/com/facebook/react/bridge/ReactTestHelper.java#L35), [ReactInstrumentationTest](https://github.com/fabriziobertoglio1987/react-native/blob/d536cf73fd8579844a17e8676aade9b2a92df0d6/ReactAndroid/src/androidTest/java/com/facebook/react/testing/ReactInstrumentationTest.java#L49))
> 8.1	task	Mock YogaConfigFactory
> 8.2	task	Read runtime stack of test (at com.facebook.yoga.YogaConfigFactory.create(YogaConfigFactory.java:12))
> 8.3	task	Read Test cases YogaLayoutDirection. They also use [YogaConfigFactory.create()](https://github.com/facebook/yoga/blob/996267dbcb1cd809f19f8d8e47d075bb982e32a8/java/tests/com/facebook/yoga/YGFlexDirectionTest.java#L29)
> 8.4	task	Write a simple test to verify correct behaviour with specific accLabel and accSpan
> 8.5	task	Use reactTextViewManager createViewInstance to create instance of ReactTextView
> 8.6	task	Test: reactTextViewManager updateState adds a TtsSpan to text

**Comment by cubuspl42** [comment]
@cortinico The first build failure when [trying to apply](https://github.com/cubuspl42/ReactNativeForkExample/pull/4/files) these docs to react-native on its current `main` is:

```
Included build /home/runner/work/ReactNativeForkExample/ReactNativeForkExample/react-native/packages/react-native-gradle-plugin has build path :react-native-gradle-plugin which is the same as included build /home/runner/work/ReactNativeForkExample/ReactNativeForkExample/ReactNativeForkExample/node_modules/react-native-gradle-plugin
```

See [this job output](https://github.com/cubuspl42/ReactNativeForkExample/actions/runs/4789110092/jobs/8516582689?pr=4)

That can [be fixed](https://github.com/cubuspl42/ReactNativeForkExample/pull/5/files) by removing this line:

```
includeBuild('../node_modules/react-native-gradle-plugin')
```

But then we get:

```
Project with path ':ReactAndroid' not found in build 'react-native'.
```

See this [job output](https://github.com/cubuspl42/ReactNativeForkExample/actions/runs/4789165189/jobs/8516701534?pr=5)

We can try to fix that [by adding](https://github.com/cubuspl42/ReactNativeForkExample/pull/6/files) `:packages:react-native` prefix.

But then we get...

```
A problem occurred starting process 'command '/home/runner/work/ReactNativeForkExample/ReactNativeForkExample/react-native/node_modules/@react-native/codegen/scripts/oss/build.sh''
```

See this [job output](https://github.com/cubuspl42/ReactNativeForkExample/actions/runs/4789216643/jobs/8516808516?pr=6)

What am I doing wrong? Or did you mean that docs are up to date, but assuming one forks a tagged version, not `main`?

**Comment by cortinico** [comment]
> What am I doing wrong? Or did you mean that docs are up to date, but assuming one forks a tagged version, not `main`?

Those docs are up to date for 0.71.
On 0.72 and on main you'll have to refer to:
- https://github.com/facebook/react-native-website/pull/3659

**Comment by cubuspl42** [comment]
@NickGerleman

> Then the other, was just making sure this logic is updated everywhere. Right now the various usages of “SetSpanOperation” are copy/pasted in a few places, but that should make it mechanical to validate the change in one place then apply to the others.

In the newest revision I'll push in the second, I de-duplicated the `SetSpanOperation` class.

> So, I’d want to see validation of the case where spans do influence each other. E.g. AbsoluteSizeSpan may influence the metrics read by CustomLineHeightSpan.

I expanded the Line Height example in RNTester with this case:

```js
<Text>
  <Text style={{backgroundColor: 'red', lineHeight: 40}}>First</Text>
  <Text style={{backgroundColor: 'yellow', lineHeight: 30, fontSize: 20}}>second</Text>
</Text>
```

On `main`, the second line height (30) seems to be ignored:

image

After my PR, the first line height (40) seems to be ignored.

image

To be honest, I have no idea what's the expected behavior.

I'll apply suggestions by @Szymon20000, so we can take advantage of the fact that on modern Android versions (maybe most?) the insertion order is also taken into consideration. This should minimize the chance of any change of behavior, even in cases like this one, where it's not obvious what's the correct behavior, or when the old behavior might have been incorrect (but, in reality, some people might relying on it).

**Comment by cubuspl42** [comment]
@NickGerleman Organizationally: I changed the base branch to `main`. Also, would you consider assigning yourself as a reviewer? I'm very grateful for your help so far and it seems that you have experience with the modified module.

**Comment by cubuspl42** [comment]
@NickGerleman Let me know what you think when you can 😃 I have some ideas on how to minimize any potential behavior changes to a minimum (even when it's not obvious which behavior is correct) by sorting the `SetSpan` objects before executing them. Do you think it's a good direction? Or should we consider some behavior undefined?

**Comment by fabriziobertoglio1987** [comment]
> @NickGerleman
> 
> > Then the other, was just making sure this logic is updated everywhere. Right now the various usages of “SetSpanOperation” are copy/pasted in a few places, but that should make it mechanical to validate the change in one place then apply to the others.
> 
> In the newest revision I'll push in the second, I de-duplicated the `SetSpanOperation` class.
> 
> > So, I’d want to see validation of the case where spans do influence each other. E.g. AbsoluteSizeSpan may influence the metrics read by CustomLineHeightSpan.
> 
> I expanded the Line Height example in RNTester with this case:
> 
> ```js
> <Text>
>   <Text style={{backgroundColor: 'red', lineHeight: 40}}>First</Text>
>   <Text style={{backgroundColor: 'yellow', lineHeight: 30, fontSize: 20}}>second</Text>
> </Text>
> ```
> 
> On `main`, the second line height (30) seems to be ignored:
> 
> image
> 
> After my PR, the first line height (40) seems to be ignored.
> 
> image
> 
> To be honest, I have no idea what's the expected behavior.
> 
> I'll apply suggestions by @Szymon20000, so we can take advantage of the fact that on modern Android versions (maybe most?) the insertion order is also taken into consideration. This should minimize the chance of any change of behavior, even in cases like this one, where it's not obvious what's the correct behavior, or when the old behavior might have been incorrect (but, in reality, some people might relying on it).

https://github.com/facebook/react-native/pull/36656#issuecomment-1519374118

https://developer.android.com/develop/ui/views/text-and-emoji/spans#paragraph-spans

>Spans that affect paragraphs
>A span can also affect text at a paragraph level, such as changing the alignment or the margin of the entire block of text. Spans that affect entire paragraphs implement [ParagraphStyle](https://developer.android.com/reference/android/text/style/ParagraphStyle). When using these spans, you must attach them to the entire paragraph, excluding the ending new line character. If you try to apply a paragraph span to something other than a whole paragraph, Android does not apply the span at all.

**Comment by fabriziobertoglio1987** [comment]
>- Paragraph Affecting spans (text separated by `\n` (new line)). A span can also affect text at a paragraph level, such as changing the alignment or the margin of the entire block of text. Spans that affect entire paragraphs implement [ParagraphStyle](https://developer.android.com/reference/android/text/style/ParagraphStyle).
>- Character Affecting spans. A span can affect text at a character level. For example, you can update character elements like background color, style, or size. Spans that affect individual characters extend the [CharacterStyle](https://developer.android.com/reference/android/text/style/CharacterStyle) class.
>- Appearance Affecting spans
>- Metric Affecting spans. Spans can also affect text metrics, such as line height and text size. All of these spans extend the [MetricAffectingSpan](https://developer.android.com/reference/android/text/style/MetricAffectingSpan) class.

#### AOSP Layout.java drawText is responsible of drawing the spans and applying the styles

An example of similar functionality on Android is the AlignmentSpan. 
A paragraph is a text separated by `\n` (new line). The AlignmentSpan changes the horizontal alignment of the paragraph.

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/text/Layout.java;l=494-500;drc=caabd536e8fea82b888902f359b33e637267966c;bpv=1;bpt=1?q=ALIGN_OPPOSITE&ss=android%2Fplatform%2Fsuperproject
>New batch of paragraph styles, collect into spans array. Compute the alignment, last alignment style wins. Reset tabStops, we'll rebuild if we encounter a line with tabs. We expect paragraph spans to be relatively infrequent, use spanEnd so that we can check less frequently.  Since paragraph styles ought to apply to entire paragraphs, we can just collect the ones present at the start of the paragraph. If spanEnd is before the end of the paragraph, that's not our problem.

Details on how the Layout.java API draws text with the AlignmentSpan.
https://github.com/facebook/react-native/issues/30375#issuecomment-782213872. 

1) Saves the paragraph alignment ([writeToParcelInternal](https://github.com/aosp-mirror/platform_frameworks_base/blob/7096deadf4a32768717e74b11eb8e6289a65976f/core/java/android/text/style/AlignmentSpan.java#L95-L98)) as `ALIGN_CENTER`, `ALIGN_OPPOSITE` or `ALIGN_NORMAL`.
3) Text and spans are drawn with Layout [#drawText][1]. The method iterates over each Nested Text and adds styles like background color, font size, and other nested Text attributes.
4) drawText iterates over each AlignmentSpan and applies their custom horizontal alignment ([source][2])
3) The [TextLine][3] API is used to measure, recycle, and draw the text.


[1]: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/text/Layout.java;l=440;drc=caabd536e8fea82b888902f359b33e637267966c?q=ALIGN_OPPOSITE&ss=android%2Fplatform%2Fsuperproject 


[2]: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/text/Layout.java;l=494-500;drc=caabd536e8fea82b888902f359b33e637267966c;bpv=1;bpt=1?q=ALIGN_OPPOSITE&ss=android%2Fplatform%2Fsuperproject


[3]: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/text/TextLine.java;l=138-161;drc=caabd536e8fea82b888902f359b33e637267966c

**Comment by cubuspl42** [comment]
@NickGerleman

While calculating the priority based on depth seems to be conceptually correct, I don't want to break some existing apps, even if they apply styles in a questionable way.

From the beginning, I wanted this change to be purely a bugfix, changing the behavior only in the case of very big text nodes, and only improving it.

As I've already spent more time on this issue than I originally intended, I decided to pivot. Instead of changing the way how the priority is calculated, I'm now suggesting a much simpler, more local fix for the overflow. It should improve the situation a lot, fixing a whole class of visual bugs.
The internal team is welcome to pick up my previous attempts to actually compute the span priority based on depth, while also considering how this might affect some styling corner cases.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/SetSpanOperation.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    // If the computed priority doesn't fit int the flags, clamp it. The effect might not be correct
+    // in 100% of cases, but doing nothing (as we did in the past) leads to totally random results.
+    final int effectivePriority = Math.max(priority, 0);
+
+    spanFlags &= ~Spannable.SPAN_PRIORITY;
+    spanFlags |= (effectivePriority << Spannable.SPAN_PRIORITY_SHIFT) & Spannable.SPAN_PRIORITY;
+
+    sb.setSpan(what, start, end, spanFlags);
```
Comment (NickGerleman) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/SetSpanOperation.java:51]: Ah, this change looks good to me 👍

**Review by NickGerleman — APPROVED** [review]

**Comment by cubuspl42** [comment]
I pushed a new revision. Fixed a typo.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java [hunk 2]
```diff
@@ -588,30 +588,4 @@ public static WritableArray measureLines(
             hyphenationFrequency);
     return FontMetricsUtil.getFontMetrics(text, layout, sTextPaintInstance, context);
   }
-
-  // TODO T31905686: This class should be private
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:592]: Thanks for doing this <3

**Review by cortinico — APPROVED** [review]

**Comment by cubuspl42** [comment]
@NickGerleman Thanks for merging this! When can I expect this to be included in a release? Does this have any chance of being included in 0.72.x, or will I have to wait till 0.73?

**Comment by NickGerleman** [comment]
You can make a pick request here: https://github.com/reactwg/react-native-releases/discussions/54

**Pull request closed without merging**